internal/provider/scheduler/farmer_bot.go

```diff
@@ -0,0 +1,169 @@
+package scheduler
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+)
+
+const (
+	FarmerBotVersionAction  = "farmerbot.farmmanager.version"
+	FarmerBotFindNodeAction = "farmerbot.nodemanager.findnode"
+	FarmerBotRMBFunction    = "execute_job"
+)
+
+type FarmerBotAction struct {
+	Guid         string        `json:"guid"`
+	TwinID       uint32        `json:"twinid"`
+	Action       string        `json:"action"`
+	Args         FarmerBotArgs `json:"args"`
+	Result       FarmerBotArgs `json:"result"`
+	State        string        `json:"state"`
+	Start        uint64        `json:"start"`
+	End          uint64        `json:"end"`
+	GracePeriod  uint32        `json:"grace_period"`
+	Error        string        `json:"error"`
+	Timeout      uint32        `json:"timeout"`
+	SourceTwinID uint32        `json:"src_twinid"`
+	SourceAction string        `json:"src_action"`
+	Dependencies []string      `json:"dependencies"`
+}
+
+type FarmerBotArgs struct {
+	Args   []Args   `json:"args"`
+	Params []Params `json:"params"`
+}
+
+type Args struct {
+	RequiredHRU  *uint64  `json:"required_hru,omitempty"`
+	RequiredSRU  *uint64  `json:"required_sru,omitempty"`
+	RequiredCRU  *uint64  `json:"required_cru,omitempty"`
+	RequiredMRU  *uint64  `json:"required_mru,omitempty"`
+	NodeExclude  []uint32 `json:"node_exclude,omitempty"`
+	Dedicated    *bool    `json:"dedicated,omitempty"`
+	PublicConfig *bool    `json:"public_config,omitempty"`
+	PublicIPs    *uint32  `json:"public_ips"`
+	Certified    *bool    `json:"certified,omitempty"`
+}
+
+type Params struct {
+	Key   string      `json:"key"`
+	Value interface{} `json:"value"`
+}
+
+func (s *Scheduler) hasFarmerBot(ctx context.Context, farmID uint32) bool {
+	_, err := s.getFarmInfo(farmID)
```

Comment:
does this function `getFarmInfo` return an error in case of no farmerbot?

Reply:
no, it only gets farm info from gridproxy. checking for farmerbot is done through an rmb call after.